### ISSUE TITLE
Fluent: WithExistingNetworkSecurityGroup not working for newly created NSG?

### ISSUE THREAD

**Issue by TimLovellSmith** [Azure/azure-sdk-for-net#3361]
I recently upgraded to 1.0.0. I had some code which I reworked since the previous version which ended up after the upgrade looking like

```
try {
    logger.Info($"Begin create NSG {nsgName} in {resourceGroup.Id}...");
                nsg = await azureSub.NetworkSecurityGroups.Define(nsgName)
                .WithRegion(resourceGroup.Region)
                .WithExistingResourceGroup(resourceGroup)
                .WithTag("foo", OwnerName)
                .DefineRule("allowAllInbound").AllowInbound().FromAnyAddress().FromAnyPort().ToAnyAddress().ToAnyPort().WithAnyProtocol().WithPriority(100).Attach()
                .DefineRule("allowAllOutbound").AllowOutbound().FromAnyAddress().FromAnyPort().ToAnyAddress().ToAnyPort().WithAnyProtocol().WithPriority(101).Attach()
                .CreateAsync();
                logger.Info($"Successfully created NSG: {nsg.Name}.");
            }
            catch (Exception e)
            {
                LogTheExceptionDetails(e, logger);
                throw;
            }

            NSGId = nsg.Id;

            // Create a VNET + subnet in the resource group
            try
            {
                logger.Info($"Begin create ARM VNET {Name} in {resourceGroup.Id} referencing NSG {NSGId}...");
                vnet = await azureSub.Networks.Define(Name)
                    .WithRegion(resourceGroup.Region)
                    .WithExistingResourceGroup(resourceGroup)
                    .WithTag("foo", OwnerName)
                    .WithAddressSpace("172.16.0.0/12")
                    .DefineSubnet("subnet").WithAddressPrefix("172.16.0.0/16").WithExistingNetworkSecurityGroup(nsg).Attach()
                    .CreateAsync();
                logger.Info($"Successfully created VNET: {vnet.Name}.");
            }
            catch (Exception e)
            {
                LogTheExceptionDetails(e, logger);
                throw;
            }
```

But it seems to have been failing due to the WithExistingNetworkSecurityGroup not working somehow in this scenario - here is the server response!

```
[Deployment: iaas202c3, Id: 277] More failure details: Response?.StatusCode=BadRequest, Content={
  "error": {
    "code": "InvalidRequestFormat",
    "message": "Cannot parse the request.",
    "details": [
      {
        "code": "MissingJsonReferenceId",
        "message": "Value for reference id is missing. Path properties.subnets[0].properties.networkSecurityGroup."
      }
    ]
  }
}
```

I don't really see why this code could fail though unless its a bug - or unless I'm not testing the code that I think I am - nsg should be initialized right, so WithExistingNetworkSecurityGroup should work right? (in the logs I also see `[Deployment: iaas202c3, Id: 277] Successfully created NSG: weur299254vnetnsg.` ...)

**Comment by martinsawicki** [comment]
intriguing indeed. I'll need to look at this.

**Comment by TimLovellSmith** [comment]
@hovsepm /cc @martinsawicki Updated to 1.1.0 but it didn't seem to help with this NSG issue in any way.

Logging shows that NSG.Id property gets initialized to '..../resourceGroups/weur793706/providers/Microsoft.Network/networkSecurityGroups/weur793706vnetnsg' by the time I call create vnet.

Was there anything else I changed when I updated to 1.0.0? Yes! I did make some changes that probably affect the timing of the calls! I can see if adding some delay in here helps.

**Comment by TimLovellSmith** [comment]
@hovsepm @martinsawicki I also tried adding the NSG to the subnet as an 'Update()' call and it doesn't work either [same response]:

```
try {
                logger.Info($"Adding the NSG reference for ARM VNET {Name} in {resourceGroup.Id} referencing NSG {NSGId}...");
                await vnet.Update().UpdateSubnet("subnet").WithExistingNetworkSecurityGroup(NSGId).Parent().ApplyAsync();
                logger.Info($"Successfully added the NSG reference for ARM VNET {Name} in {resourceGroup.Id} referencing NSG {NSGId}...");
            }
```

**Comment by TimLovellSmith** [comment]
+ @anuchandy I notice the symptom resembles #3359, however your fix only address VMs right?

**Comment by anuchandy** [comment]
@TimLovellSmith issue looks similar, SDK is not serializing the id of the NSG in this case.

@martinsawicki Looking at the corresponding java [type](https://github.com/Azure/azure-sdk-for-java/blob/master/azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/SubnetInner.java#L31), NSG property is a SubResource. In case of C# it is a Resource hence the ID will not get serialized on PUT.

**Comment by hovsepm** [comment]
fix published to nuget. Please upgrade to version 1.1.1.

**Comment by hovsepm** [comment]
Still fails on Update

**Comment by hrishi18pathak** [comment]
Actually it fails on the create
We also get a similar failure while creating loadbalancer
Code:

// Create the LoadBalancer, and all the port-forwarding rules we will need to support
            // a cluster of VMs.
            // Wait for dependent tasks before continuing
            Func<Task<ILoadBalancer>> createLoadBalancer = async () => {
                var lbName = GetLoadBalancerName();
                Console.WriteLine($"Begin create LOAD BALANCER {lbName}...");
                var loadBalancerDef = azureSub.LoadBalancers.Define(lbName)
                    .WithRegion(resourceGroup.Region)
                    .WithExistingResourceGroup(resourceGroup.Name)
                    .DefinePublicFrontend("frontend")
                    .WithExistingPublicIpAddress(ipAddress)
                    .Attach()
                    .DefineBackend("backend").Attach()
                    .DefineTcpProbe("probe10225").WithPort(10225).WithIntervalInSeconds(5).WithNumberOfProbes(3).Attach()
                    .DefineHttpProbe("redisprobe").WithRequestPath("/api/redisprobe").WithPort(8500).WithIntervalInSeconds(5).WithNumberOfProbes(3).Attach()
                    //.DefineHttpProbe("redismasterprobe").WithRequestPath("/api/redismasterprobe").WithPort(8500).WithIntervalInSeconds(5).WithNumberOfProbes(3).Attach()
                    .DefineLoadBalancingRule("tcp6379")
                        .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend").WithFrontendPort(6379) // 'RedisPublic'
                        .WithProbe("redisprobe")
                        .WithBackend("backend").WithBackendPort(6379)
                        .WithIdleTimeoutInMinutes(20)
                        .Attach()
                    .DefineLoadBalancingRule("tcp6380")
                        .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend").WithFrontendPort(6380) // 'RedisSecure'
                        .WithProbe("redisprobe")
                        .WithBackend("backend").WithBackendPort(6380)
                        .WithIdleTimeoutInMinutes(20)
                        .Attach()
                    .DefineLoadBalancingRule("tcp10225")
                        .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend")
                        .WithFrontendPort(10225) // 'PublicHttpPort'
                        .WithProbe("probe10225")
                        .WithBackend("backend").WithBackendPort(10225)
                        .WithIdleTimeoutInMinutes(20)
                        .Attach()
                    .DefineInboundNatPool("pool330XX")
                    .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend")
                        .WithFrontendPortRange(33000, 33019)
                        .WithBackendPort(3389)
                        .Attach()
                     .DefineInboundNatPool("pool160XX")
                    .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend")
                        .WithFrontendPortRange(16000, 16019)
                        .WithBackendPort(10126)
                        .Attach()
                        .DefineInboundNatPool("pool130XX")
                    .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend")
                        .WithFrontendPortRange(13000, 13019)
                        .WithBackendPort(10227)
                        .Attach()
                        .DefineInboundNatPool("pool150XX")
                    .WithProtocol(TransportProtocol.Tcp)
                        .WithFrontend("frontend")
                        .WithFrontendPortRange(15000, 15019)
                        .WithBackendPort(10221)
                        .Attach();

                var ret = await loadBalancerDef
                    .WithTag("redis", OwnerName)
                    .CreateAsync(token);

Error: 
Value for reference id is missing. Path properties.frontendIPConfigurations[0].properties.publicIPAddress. 

The IP address is created and has Id: 
/subscriptions/c89fae2d-0f46-42eb-821b-5f0d36f4444f/resourceGroups/hpvmssrg/providers/Microsoft.Network/publicIPAddresses/hpvmsstestip

**Comment by hrishi18pathak** [comment]
This is the stack trace:
   at Microsoft.Azure.Management.Network.Fluent.LoadBalancersOperations.<BeginCreateOrUpdateWithHttpMessagesAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Management.Network.Fluent.LoadBalancersOperations.<CreateOrUpdateWithHttpMessagesAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Management.Network.Fluent.LoadBalancersOperationsExtensions.<CreateOrUpdateAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Management.Resource.Fluent.Core.GroupableParentResource`8.<CreateResourceAsync>d__5.MoveNext()

**Comment by hovsepm** [comment]
fixed in [v1.1.2](https://www.nuget.org/packages/Microsoft.Azure.Management.Fluent/1.1.2)
